### PR TITLE
Only generate BF16 conversion builtins when considered supported

### LINKED ISSUE #1847: [PyTorch Upstream][MTL]  bf conversion instruction not supported
# Bug Description

When running the BF16 related huggingface models on **MTL**, it complains about the following:

```Bash
L0 build module failed. Log:
error: bf conversion instruction not supported!
in kernel: 'triton_poi_fused__to_copy_2'
error: backend compiler failed build.

Traceback (most recent call last):
  File "/home/gta/actions-runner/actions-runner/_work/torch-xpu-ops/repro.py", line 34, in <module>
    async_compile.wait(globals())
  File "/opt/conda/envs/e2e_ci/lib/python3.10/site-packages/torch/_inductor/async_compile.py", line 261, in wait
    scope[key] = result.result()
  File "/opt/conda/envs/e2e_ci/lib/python3.10/site-packages/torch/_inductor/codecache.py", line 3745, in result
    self.kernel.precompile()
  File "/opt/conda/envs/e2e_ci/lib/python3.10/site-packages/torch/_inductor/runtime/triton_heuristics.py", line 234, in precompile
    compiled_binary, launcher = self._precompile_config(
  File "/opt/conda/envs/e2e_ci/lib/python3.10/site-packages/torch/_inductor/runtime/triton_heuristics.py", line 442, in _precompile_config
    binary._init_handles()
  File "/opt/conda/envs/e2e_ci/lib/python3.10/site-packages/triton/compiler/compiler.py", line 376, in _init_handles
    self.module, self.function, self.n_regs, self.n_spills = driver.active.utils.load_binary(
RuntimeError: Triton Error [ZE]: 0x70000004
```

This issue affects 44/46 models in Huggingface bf16 related inference. 

# Reproducer

```Python
import torch
from torch._inductor.async_compile import AsyncCompile
async_compile = AsyncCompile()


triton_poi_fused__to_copy_2 = async_compile.triton('triton_poi_fused__to_copy_2', '''
import triton
import triton.language as tl
from triton.compiler.compiler import AttrsDescriptor

from torch._inductor.runtime import triton_helpers, triton_heuristics
from torch._inductor.runtime.triton_helpers import libdevice, math as tl_math
from torch._inductor.runtime.hints import AutotuneHint, ReductionHint, TileHint, instance_descriptor, DeviceProperties

@triton_heuristics.pointwise(
    size_hints=[1048576], 
    filename=__file__,
    triton_meta={'signature': {0: '*fp32', 1: '*bf16', 2: 'i32'}, 'device': DeviceProperties(type='xpu', index=0, cc={'driver_version': '1.3.29735', 'gpu_eu_count': 128, 'gpu_subslice_count': 8, 'has_atomic64': True, 'has_fp16': True, 'has_fp64': True, 'max_compute_units': 128, 'max_num_sub_groups': 128, 'max_work_group_size': 1024, 'name': 'Intel(R) Arc(TM) Graphics', 'platform_name': 'Intel(R) Level-Zero', 'sub_group_sizes': [8, 16, 32], 'total_memory': 14793265152, 'type': 'gpu', 'vendor': 'Intel(R) Corporation', 'version': '1.3'}, major=None, regs_per_multiprocessor=None, max_threads_per_multi_processor=None, multi_processor_count=None), 'constants': {}, 'configs': [AttrsDescriptor(divisible_by_16=(0, 1, 2), equal_to_1=())]},
    inductor_meta={'autotune_hints': set(), 'kernel_name': 'triton_poi_fused__to_copy_2', 'mutated_arg_names': [], 'no_x_dim': False, 'num_load': 1, 'num_reduction': 0, 'backend_hash': '0530A25570032EB138057C3031779D0C791D2D9CB2D7D53F7BDC87FB6096590F', 'are_deterministic_algorithms_enabled': True, 'assert_indirect_indexing': True, 'autotune_local_cache': True, 'autotune_pointwise': True, 'autotune_remote_cache': None, 'force_disable_caches': False, 'dynamic_scale_rblock': True, 'max_autotune': False, 'max_autotune_pointwise': False, 'min_split_scan_rblock': 256, 'spill_threshold': 16, 'store_cubin': False},
    min_elem_per_thread=0
)
@triton.jit
def triton_poi_fused__to_copy_2(in_ptr0, out_ptr0, xnumel, XBLOCK : tl.constexpr):
    xnumel = 589824
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:]
    xmask = tl.full([XBLOCK], True, tl.int1)
    x0 = xindex
    tmp0 = tl.load(in_ptr0 + (x0), None)
    tmp1 = tmp0.to(tl.float32)
    tl.store(out_ptr0 + (x0), tmp1, None)
''', device_str='xpu')

async_compile.wait(globals())
```

# Env Building

```Bash
git clone https://github.com/pytorch/pytorch
# Then install driver/oneapi following https://github.com/pytorch/pytorch?tab=readme-ov-file#intel-gpu-support
# Install the same dependencies as https://github.com/pytorch/pytorch?tab=readme-ov-file#install-dependencies

# Source oneAPI before building pytorch
source /opt/intel/oneapi/setvars.sh

# build PyTorch
cd pytorch
python setup.py install

# Install Triton
pip install --pre pytorch-triton-xpu --index-url https://download.pytorch.org/whl/nightly/xpu
```

# Env Detail

```Bash
Triton Version: pytorch-triton-xpu-3.0.0+1b2f15840e

PyTorch version: 2.5.0a0+git7d5462b
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.30.2
Libc version: glibc-2.35

Python version: 3.10.14 | packaged by conda-forge | (main, Mar 20 2024, 12:45:18) [GCC 12.3.0] (64-bit runtime)
Python platform: Linux-6.6.7-060607-generic-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True


Versions of relevant libraries:
[pip3] bert_pytorch==0.0.1a4
[pip3] functorch==1.14.0a0+b71aa0b
[pip3] mypy==1.11.1
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.4
[pip3] onnx==1.16.2
[pip3] optree==0.12.1
[pip3] pytorch-labs-segment-anything-fast==0.2
[pip3] torch==2.5.0a0+git7d5462b
[pip3] torch_geometric==2.4.0
[pip3] torchao==0.4.0
[pip3] torchaudio==2.4.0a0+b3f6f51
[pip3] torchmetrics==1.0.3
[pip3] torchmultimodal==0.1.0b0
[pip3] torchrec==0.9.0.dev20240811+cpu
[pip3] torchvision==0.19.0a0+d23a6e1
[pip3] triton==3.0.0
[conda] bert-pytorch              0.0.1a4                   dev_0    <develop>
[conda] functorch                 1.14.0a0+b71aa0b          pypi_0    pypi
[conda] mkl-include               2024.2.1                 pypi_0    pypi
[conda] mkl-static                2024.2.1                 pypi_0    pypi
[conda] numpy                     1.26.4                   pypi_0    pypi
[conda] optree                    0.12.1                   pypi_0    pypi
[conda] pytorch-labs-segment-anything-fast 0.2                      pypi_0    pypi
[conda] torch                     2.5.0a0+git7d5462b          pypi_0    pypi
[conda] torch-geometric           2.4.0                    pypi_0    pypi
[conda] torchao                   0.4.0                    pypi_0    pypi
[conda] torchaudio                2.4.0a0+b3f6f51          pypi_0    pypi
[conda] torchmetrics              1.0.3                    pypi_0    pypi
[conda] torchmultimodal           0.1.0b0                  pypi_0    pypi
[conda] torchrec                  0.9.0.dev20240811+cpu          pypi_0    pypi
[conda] torchvision               0.19.0a0+d23a6e1          pypi_0    pypi
[conda] triton                    3.0.0                    pypi_0    pypi
```

### PR DESCRIPTION
**Pull request by whitneywhtsang**
Fixes #1847

### CODE REVIEW

**Comment by whitneywhtsang** [comment]
@ZzEeKkAa Please help check if `has_bf16_conversion` is false on MTL.

File: third_party/intel/lib/TritonIntelGPUToLLVM/BF16Casts.cpp [hunk 1]
```diff
@@ -112,6 +123,13 @@ Value convertFp32ToBf16(Location loc, ConversionPatternRewriter &rewriter,
   auto exponent_not_all1s = icmp_ne(check_exponent, i32_val(0));
   auto exponent_all1s = icmp_eq(check_exponent, i32_val(0));
   Value rounded = as_uint32;
+  if (rounding == RoundingMode::RTNE) {
+    rounded =
+        add(i32_ty, i32_val(0x7fff),
+            and_(i32_ty, lshr(i32_ty, as_uint32, i32_val(16)), i32_val(1)));
+    rounded = add(i32_ty, rounded, as_uint32);
+    rounded = select(exponent_not_all1s, rounded, as_uint32);
+  }
```
Comment (victor-eds) [inline third_party/intel/lib/TritonIntelGPUToLLVM/BF16Casts.cpp:132]: Complex stuff

**Review by victor-eds — COMMENTED** [review]
LGTM. Just 1 question

**Comment by sommerlukas** [comment]
Is there any overlap of this and #1285?

**Comment by victor-eds** [comment]
> Is there any overlap of this and #1285?

This improves (fixes?) the current approach. My suggestion there is to leverage an existing MLIR utility to do the conversion as an additional step, it's basically refactoring, would take more effort.

File: third_party/intel/backend/compiler.py [hunk 2]
```diff
@@ -137,6 +137,7 @@ def parse_target(self, tgt_prop) -> dict:
         dev_prop['has_subgroup_matrix_multiply_accumulate_tensor_float32'] = tgt_prop.get(
             'has_subgroup_matrix_multiply_accumulate_tensor_float32', False)
         dev_prop['has_subgroup_2d_block_io'] = tgt_prop.get('has_subgroup_2d_block_io', False)
+        dev_prop['has_bf16_conversion'] = tgt_prop.get('has_bf16_conversion', False)
```
Comment (vlad-penkin) [inline third_party/intel/backend/compiler.py:140]: `has_bf16_conversion` property value is the derivative of the device architecture. I suggest to set it explicitly to False for MTL only.

**Review by vlad-penkin — CHANGES_REQUESTED** [review]
@whitneywhtsang 

FYI: PyTorch Team is asking to backport this change to the PyTorch 2.4 / Triton 3.0 branch. 

Please take it into the consideration, this may change the way we structure PR's to implement this change.

File: third_party/intel/lib/TritonIntelGPUToLLVM/BF16Casts.cpp [hunk 3]
```diff
@@ -74,22 +76,31 @@ namespace mlir::triton::intel {
...
+        triton::gpu::intel::lookupOrCreateSPIRVFn(moduleOp, name, inTy, outTy);
+    auto call = triton::gpu::intel::createSPIRVBuiltinCall(
+        loc, rewriter, ext_func, bitcast(v, inTy).getResult());
+    return call.getResult();
+  }
+
+  auto as_int16 = bitcast(v, i16_ty);
+  auto as_int32 = zext(i32_ty, as_int16);
+  auto shifted = shl(i32_ty, as_int32, i32_val(16));
+  return (bitcast(shifted, f32_ty));
```
[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply third_party/intel/lib/TritonIntelGPUToLLVM/BF16Casts.cpp:95]: The reported issue is with scalar type, that's why keeping vector support as an assertion.

File: third_party/intel/backend/compiler.py (hunk 2 above)
Comment (etiotto) [reply third_party/intel/backend/compiler.py:140]: It is the responsibility of PyTorch to set these flags correctly based on the target GPU.

**Comment by whitneywhtsang** [comment]
> @whitneywhtsang
> 
> FYI: PyTorch Team is asking to backport this change to the PyTorch 2.4 / Triton 3.0 branch.
> 
> Please take it into the consideration, this may change the way we structure PR's to implement this change.

This change can be backported to the Triton 3.0 branch.

File: third_party/intel/lib/TritonIntelGPUToLLVM/BF16Casts.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply third_party/intel/lib/TritonIntelGPUToLLVM/BF16Casts.cpp:95]: Let's do it as an additional change if PyTorch team require it for vector type?

[the remark this answers is not in the document]
Comment (victor-eds) [reply third_party/intel/lib/TritonIntelGPUToLLVM/BF16Casts.cpp:95]: Nice

**Review by victor-eds — APPROVED** [review]

**Comment by ZzEeKkAa** [comment]
> @ZzEeKkAa Please help check if `has_bf16_conversion` is false on MTL.

I need to request machine. Will let you know

File: third_party/intel/backend/compiler.py (hunk 2 above)
Comment (whitneywhtsang) [reply third_party/intel/backend/compiler.py]: Changed default to true as suggested.

File: third_party/intel/backend/compiler.py [hunk 4]
```diff
@@ -185,7 +186,7 @@ def make_ttgir(mod, metadata, opt, properties):
         intel.passes.ttgpuir.add_triton_annotate_module(pm, min(properties["sub_group_sizes"]),
                                                         properties["has_subgroup_2d_block_io"],
                                                         properties["has_subgroup_matrix_multiply_accumulate"],
-                                                        opt.threads_per_warp)
+                                                        properties["has_bf16_conversion"], opt.threads_per_warp)
```
Comment (vlad-penkin) [inline third_party/intel/backend/compiler.py:189]: For consistency please move `opt.threads_per_warp)` to the next line.

**Review by vlad-penkin — CHANGES_REQUESTED** [review]

File: third_party/intel/backend/compiler.py (hunk 4 above)
Comment (whitneywhtsang) [reply third_party/intel/backend/compiler.py:189]: it is not controlled by me, pre_commit (clang-format) decided to put it this way.

**Review by vlad-penkin — APPROVED** [review]
LGTM

**Review by etiotto — APPROVED** [review]

**Comment by ZzEeKkAa** [comment]
I've run https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_compiler_opencl.asciidoc#querying-supported-features-and-extensions on MTL, and this is the output:
```
Device supports online compilation of OpenCL C kernels
Device supports online compilation with OpenCL C 3.0
Device supports online compilation with 64-bit FP in OpenCL C
Device supports online compilation with the OpenCL full profile
```

So I would expect bloat16 return false.

**Comment by ZzEeKkAa** [comment]
And finally, I have the output:

```
python -c "import torch, json; print(json.dumps(torch.xpu.get_device_capability()))" | jq
[W815 20:57:42.025124416 OperatorEntry.cpp:155] Warning: Warning only once for all operators,  other operators may also be overridden.
  Overriding a previously registered kernel for the same operator and the same dispatch key
  operator: aten::bitwise_left_shift.Tensor_out(Tensor self, Tensor other, *, Tensor(a!) out) -> Tensor(a!)
    registered at /home/yevhenii/Projects/pytorch/build/aten/src/ATen/RegisterSchema.cpp:6
  dispatch key: XPU
  previous kernel: registered at /home/yevhenii/Projects/pytorch/build/aten/src/ATen/RegisterCPU.cpp:30476
       new kernel: registered at /home/yevhenii/Projects/pytorch/build/aten/src/ATen/xpu/RegisterXPU.cpp:7756 (function operator())
{
  "driver_version": "1.3.29735",
  "gpu_eu_count": 128,
  "gpu_subslice_count": 8,
  "has_atomic64": true,
  "has_bf16_conversion": false,
  "has_fp16": true,
  "has_fp64": true,
  "has_subgroup_2d_block_io": false,
  "has_subgroup_matrix_multiply_accumulate": false,
  "has_subgroup_matrix_multiply_accumulate_tensor_float32": false,
  "max_compute_units": 128,
  "max_num_sub_groups": 128,
  "max_work_group_size": 1024,
  "name": "Intel(R) Arc(TM) Graphics",
  "platform_name": "Intel(R) Level-Zero",
  "sub_group_sizes": [
    8,
    16,
    32
  ],
  "total_memory": 7158026240,
  "type": "gpu",
  "vendor": "Intel(R) Corporation",
  "version": "1.3"
}
```

**Pull request merged**